File: AIServerAndroid/app/src/main/kotlin/com/gemma4/aiserver/server/routes/CatalogRoutes.kt
package com.gemma4.aiserver.server.routes

import com.gemma4.aiserver.catalog.ModelCatalog
import com.gemma4.aiserver.download.DownloadRepository
import com.gemma4.aiserver.inference.InferenceRepository
import com.gemma4.aiserver.model.CatalogResponse
import io.ktor.server.application.call
import io.ktor.server.response.respond
import io.ktor.server.routing.Route
import io.ktor.server.routing.get

fun Route.catalogRoutes(
    inference: InferenceRepository,
    download: DownloadRepository,
) {
    /**
     * Returns the Android model catalog with live downloaded/loaded state
     * so AIClient can render the model picker without additional round-trips.
     */
    get("/v1/catalog") {
        val models = ModelCatalog.androidModels.map { model ->
            model.copy(
                downloaded = download.isDownloaded(model.id),
                loaded = inference.loadedModelId == model.id,
            )
        }
        call.respond(CatalogResponse(models = models))
    }
}

File: AIServeriOS/AIServeriOS/Server/Routes/CatalogRoutes.swift
import Foundation

func registerCatalogRoutes(router: Router) {
    router.add("GET", "/v1/catalog") { _ in
        let download  = DownloadService.shared
        let inference = InferenceService.shared
        let loadedId  = inference.loadedModelId

        let models = ModelCatalog.all.map { info -> ModelInfo in
            var m = info
            m.downloaded = download.isDownloaded(m)
            m.loaded     = (m.id == loadedId)
            return m
        }
        return HttpResponse.json(CatalogResponse(models: models))
    }
}
